lib/authentication.js

```diff
@@ -0,0 +1,247 @@
+      'uppercase chars',
+    options: { number: 1 }
+  },
+  MIN_NUMBERS: {
+    test: (password, options) => {
+      const NUMBERS_RANGE = [[49, 57]];
+      return stringIncludesChars(password, NUMBERS_RANGE, options.number);
+    },
+    hint: options => `The password must contain more than ${options.number} ` +
+      'numbers',
+    options: { number: 1 }
+  },
+  MIN_SPECIAL_CHARS: {
+    test: (password, options) => {
+      const SPECIAL_CHARS_RANGE = [[32, 47], [58, 64], [91, 96], [123, 126]];
+      // https://www.owasp.org/index.php/Password_special_characters
+      return stringIncludesChars(password, SPECIAL_CHARS_RANGE, options.number);
+    },
+    hint: options => `The password must contain more than ${options.number} ` +
+      'special chars: !"#$%&\'()*+,-./`:;<=>?@[\\]^_{|}~', // eslint-disable-line
+    options: { number: 1 }
+  }
+};
+
+const loginTests = {
+  MIN_LENGTH: {
+    test: (login, options) => login.length >= options.minLength,
+    hint: options => `The login must be longer than ${options.minLength} chars`,
+    options: { minLength: 6 }
+  },
+  MAX_LENGTH: {
+    test: (login, options) => login.length <= options.maxLength,
+    hint: options => `The login must be less than ${options.maxLength} chars`,
+    options: { maxLength: 50 }
+  },
+  IS_EMAIL: {
+    test: login => {
+      const EMAIL_REGEXP = new RegExp(
+        '^[a-z0-9!#$%&\'*+/=?^_`{|}~-]+(?:.[a-z0-9!#$%&\'*+/=?^_`{|}~-]+)*@(' +
+        '?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$'
+      );
+      const MAX_DOMAIN_LENGTH = 255;
+      const MAX_LOCAL_PART_LENGTH = 64;
+
+      if (login.includes('@')) {
+        const [localPart, domain] = login.split('@');
+        return domain.length <= MAX_DOMAIN_LENGTH &&
+          localPart.length <= MAX_LOCAL_PART_LENGTH &&
+          EMAIL_REGEXP.test(login);
+      }
+    },
+    hint: () => 'Login must be email address'
+  }
+};
+
+const loginPasswordTests = {
+  LOGIN_INCLUDES_PASSWORD: {
+    test: (login, password) => !login.includes(password),
+    hint: () => 'The login can not contain a password'
+  },
+  PASSWORD_INCLUDES_LOGIN: {
+    test: (login, password) => !password.includes(login),
+    hint: () => 'The password can not contain a login'
+  }
+};
+
+class AuthenticationStrength {
+  constructor(
+    valid, // Boolean
+    hints, // Array, [{ required: Array, optional: Array }, ...]
+    compliance // Number, percent of best test group
+  ) {
+    this.valid = valid;
+    this.hints = hints;
+    this.strength = !valid ? 'Not valid' :
+      compliance < 0.25 ? 'Very weak' :
+        compliance < 0.4 ? 'Weak' :
+          compliance < 0.6 ? 'Good' :
+            compliance < 0.8 ? 'Strong' : 'Very strong';
+  }
+}
+
+const makeTest = (
+  // Function that checks the arguments on a test suite
+  tests, // Array, array of password/login tests
+  testGroups, // Array, [{ required: [], optional: [] }]
```

Comment:
That's what I've been concerned about. I assume this is due to the way OWASP provides their recommendations (password is 'minLen+symbols+...' OR 'isPhrase') but IMO it is better to make user manually check that if they need to support multiple 'acceptable' conditions (this would be as simple as calling this function twice with different testGroup) instead of introducing this complexity to the implementation.

Reply:
@lundibundi I can't seem to find the recommendation you mentioned in the article you provided, can you quote it, please?

Reply:
> Longer passwords provide a greater combination of characters and consequently make it more difficult for an attacker to guess.
Minimum length of the passwords should be enforced by the application.
Passwords shorter than 10 characters are considered to be weak (NIST SP800-132).
While minimum length enforcement may cause problems with memorizing passwords among some users, applications should encourage them to set passphrases (sentences or combination of words) that can be much longer than typical passwords and yet much easier to remember.
Maximum password length should not be set too low, as it will prevent users from creating passphrases. Typical maximum length is 128 characters.
Passphrases shorter than 20 characters are usually considered weak if they only consist of lower case Latin characters.

While not an explicit reccomendation this assumes that if user sets a password (not a passphrase) then minLength should be at least 10 while if he uses a passphrase it should be at least 20.
I think @bugagashenkj can clarify better the reasons behind such implementation.

Though as I mentioned I'd prefer to not allow multiple testGroups and instead make user do it manually if they need it.

Reply:
From the recommendations of owasp:
1. As already said the password is valid, if it is longer than 10 characters, it does not have duplicate characters... OR if it is longer of 20 characters(passphrase).
2. Login is valid if it is longer than minLength... OR if it is email.
But beyond that, it can be any custom test configuration.